core/WASM_core.md

```diff
@@ -0,0 +1,350 @@
+
+## Implementation Strategy
+
+### How much, if any WASI to use?
+
+The more we can lean on WASI, the less we have to expand the Core<-->Lang interface. That's a good
+thing in terms of both immediate work and long-term maintenance. There are some problems though:
+
+* Assuming we need to use `wazero` for non-CGo support, it doesn't appear to be able to handle
+  network calls. We'll test this, though.
+* Calling back into lang gives us opportunities for more user customization, which can be nice.
+  Especially when it comes to the metrics and tracing stuff. gRPC interceptors are also a potential
+  plus.
+
+My inclination for now is to at least *investigate* how much we can get out of WASI. Particularly
+with respect to running from Go.
+
+Unfortunately, my assumption is that we will need to callback into Lang for all the usecases
+mentioned at the top of the doc. The only sys interface things it appears we'll be able to reliably
+use are clocks (which is good, that does come up) and possibly file I/O, which currently does not
+happen in Core (and probably never should) except for in the test server usecase.
+
+
+### Core API changes
+Operating under the assumption that all the major syscalls will need to be handled through callbacks
+into lang. In WASM, calling functions in the host is supported via `import`ing those host functions.
+Ultimately these functions may be bundled together into a trait or something on the Core side, but
+this doc will focus on the individual functions.
+
+Here are the functions I think we'll need to add to the Core<-->Lang interface, beyond the existing
+ones that lang already calls. They're defined as they would appear in Rust, after being imported.
+This is definitely a rough draft, as it will be influenced by runtime selection and whatever
+practical problems we run into along the way.
+
+```rust
+// In reality, everything in these functions is going to need to be serialized and shuttled back
+// and forth as bytes. That makes for uninteresting signatures, though, so instead I show what
+// those signatures would look like if we were using deserialized types directly. The exact details
+// of how that serialization happens are intentionally not included here.
+//
+// WASM has an `ExternRef` type which is a pointer to a host-owned opaque object.
+
+// For all the types which wrap an `ExternRef`, the implementations on those types which take
+// `self` will be implemented as calls to a host-defined function which looks like:
+fn func_name(receiver: ExternRef, arg: ArgType) -> ResultType;
+// There may be a more clever way to manage some of this by using `Table` and `Func` types along with
+// `ExternRef` but I need to play around more to determine that.
+
+// -------------- Grpc -----------------
+
+struct GrpcClientRef {
+    /// Points to the host's gRPC client
+    host_ref: ExternRef,
+}
+
+struct GrpcRequest {
+    uri: String,
+    body: Vec<u8>,
+    metadata: Metadata,
```

Comment:
What about deadline?

Reply:
The `grpc-timeout` header specifies the deadline per https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md

Reply:
Lang impls would have to pop from the metadata since both Go and JS specify the deadline separately.
That's fine but not optimal from the lang perspective.

Reply:
Yep we can adjust as needed once we start digging in. Closing this thread since it effectively is the same as https://github.com/temporalio/proposals/pull/78#discussion_r1136328810